### PR TITLE
Make scripting work with segmented addresses

### PR DESCRIPTION
**Pull request by ptomin**
- Create sample project with segmented addresses.
- Modify `RekoProgramAPI` to accept `Reko.Core.Address` instead of
integer (linear address).
- Modify Python API to allow iteration over procedures found by scanner.
- Add possibility to read segmented memory slices.

### CODE REVIEW

File: src/Scripts/Python/RekoProgramAPI.cs [hunk 1]
```diff
@@ -160,71 +169,78 @@ public void SetUserProcedure(ulong linearAddress, string decl)
...
+        /// Returns true if there is user procedure definition at specified
+        /// address; otherwise, false
+        /// </summary>
+        private bool TryGetUserProcedure(Address addr, out UserProcedure userProc)
         {
-            var addr = Addr(linearAddress);
-            return program.User.Procedures[addr];
+            if (program.User.Procedures.TryGetValue(addr, out userProc))
+                return true;
+            var proc = program.Procedures[addr];
```
Comment (uxmal) [inline src/Scripts/Python/RekoProgramAPI.cs:241]: I would have expected the out parameter `userProc` to be null here, and have the caller make the `UserProcedure`.  Also, calling `program.Procedures[addr]` will crash if no procedure exists at address `addr`; is this the intention?

File: src/Core/Program.cs [hunk 2]
```diff
@@ -686,7 +686,14 @@ public UserGlobal ModifyUserGlobal(IProcessorArchitecture arch, Address address,
                 this.ImageMap.AddItemWithSize(address, item);
             else
                 this.ImageMap.AddItem(address, item);
-
+            //$BUGBUG: what about x86 segmented binaries?
```
Comment (uxmal) [inline src/Core/Program.cs:689]: If `address` has a selector, then you can look for the segment in `SegmentMap`.

File: src/Scripts/Python/RekoProgramAPI.cs [hunk 3]
```diff
@@ -42,97 +42,106 @@ public RekoProgramAPI(Program program)
...
+        public void SetUserGlobal(Address addr, string decl)
         {
-            var addr = Addr(linearAddress);
             var usb = new UserSignatureBuilder(program);
             var global = usb.ParseGlobalDeclaration(decl);
             var name = global?.Name;
             var dataType = global?.DataType;
             if (name is null || dataType is null)
                 throw new ArgumentException(
                     $"Failed to parse global variable declaration: '{decl}'");
```
Comment (uxmal) [inline src/Scripts/Python/RekoProgramAPI.cs:117]: Please add a '.' to the end of exception messages, for consistency.

File: src/Scripts/Python/RekoProgramAPI.cs [hunk 4]
```diff
@@ -160,71 +169,78 @@ public void SetUserProcedure(ulong linearAddress, string decl)
...
             while (rdr.Offset - offStart < length)
             {
                 yield return reader(rdr);
             }
         }
 
-        private UserProcedure UserProcedure(ulong linearAddress)
+        private UserProcedure UserProcedure(Address addr)
+        {
+            TryGetUserProcedure(addr, out var userProc);
```
Comment (uxmal) [inline src/Scripts/Python/RekoProgramAPI.cs:220]: Isn't this a bit of a lie? True, we have a userproc, but it's not registered in the `program`.

**Review by uxmal — COMMENTED** [review]
I've added some comments.

File: src/Scripts/Python/RekoProgramAPI.cs (hunk 3 above)
Comment (ptomin) [reply src/Scripts/Python/RekoProgramAPI.cs]: Done in 608181f

File: src/Scripts/Python/RekoProgramAPI.cs (hunk 1 above)
Comment (ptomin) [reply src/Scripts/Python/RekoProgramAPI.cs]: The out parameter userProc to be null after 2408f40.
> calling program.Procedures[addr] will crash if no procedure exists at address addr; is this the intention

Yes, this is the intention. I have done explicit throwing of exception at  2408f40 to make it clear.

File: src/Scripts/Python/RekoProgramAPI.cs (hunk 4 above)
Comment (ptomin) [reply src/Scripts/Python/RekoProgramAPI.cs]: Agree. I tried to make it clear at 2408f40

File: src/Core/Program.cs (hunk 2 above)
Comment (ptomin) [reply src/Core/Program.cs:689]: I tried to add it to `ImageSegment.Fields` but it is not shown at output file. Is there support for user-defined segmented global variables?

Comment (uxmal) [reply src/Core/Program.cs:689]: We should add unit tests to confirm this, both at the assembler output stage and during type inference.

Comment (ptomin) [reply src/Core/Program.cs:689]: Should I attach such tests to this PR?

Comment (ptomin) [reply src/Core/Program.cs:689]: I can try to create test at `https://github.com/uxmal/reko/blob/7efb910c1420c51de9d0364464cd87601a8a8554/src/UnitTests/Core/Output/SegmentFilePolicyTests.cs`.

Comment (ptomin) [reply src/Core/Program.cs:689]: I have added test (failed) for `SegmentFilePolicy` at 78f9de3

Comment (ptomin) [reply src/Core/Program.cs:689]: Finally, I have found solution. The fix is coming.

Comment (ptomin) [reply src/Core/Program.cs]: Done in 162c14b and 8bc0331

**Comment by ptomin** [comment]
John, I have done requested changes. Could it be merged?

**Comment by uxmal** [comment]
Nice work, Pavel.

**Pull request merged**